localization/ndt_scan_matcher/src/map_update_module.cpp

```diff
@@ -0,0 +1,302 @@
+  (*state_ptr_)["state"] = "Sleeping";
+  res->success = true;
+  res->pose_with_covariance.pose.covariance = req->pose_with_covariance.pose.covariance;
+
+  last_update_position_ = res->pose_with_covariance.pose.pose.position;
+}
+
+void MapUpdateModule::callback_ekf_odom(nav_msgs::msg::Odometry::ConstSharedPtr odom_ptr)
+{
+  current_position_ = odom_ptr->pose.pose.position;
+
+  if (last_update_position_ == std::nullopt) {
+    return;
+  }
+  double distance = norm_xy(current_position_.value(), last_update_position_.value());
+  if (distance + lidar_radius_ > dynamic_map_loading_map_radius_) {
+    RCLCPP_ERROR_STREAM_THROTTLE(logger_, *clock_, 1, "Dynamic map loading is not keeping up.");
+  }
+}
+
+void MapUpdateModule::map_update_timer_callback()
+{
+  if (current_position_ == std::nullopt) {
+    RCLCPP_ERROR_STREAM_THROTTLE(
+      logger_, *clock_, 1,
+      "Cannot find the reference position for map update. Please check if the EKF odometry is "
+      "provided to NDT.");
+    return;
+  }
+  if (last_update_position_ == std::nullopt) return;
+
+  // continue only if we should update the map
+  if (should_update_map(current_position_.value())) {
+    RCLCPP_INFO(logger_, "Start updating NDT map (timer_callback)");
+    update_map(current_position_.value());
+    last_update_position_ = current_position_;
+  }
+}
+
+bool MapUpdateModule::should_update_map(const geometry_msgs::msg::Point & position) const
+{
+  if (last_update_position_ == std::nullopt) return false;
+  double distance = norm_xy(position, last_update_position_.value());
+  return distance > dynamic_map_loading_update_distance_;
+}
+
+void MapUpdateModule::update_map(const geometry_msgs::msg::Point & position)
+{
+  auto request = std::make_shared<autoware_map_msgs::srv::GetDifferentialPointCloudMap::Request>();
+  request->area.center = position;
+  request->area.radius = dynamic_map_loading_map_radius_;
+  request->cached_ids = ndt_ptr_->getCurrentMapIDs();
+
+  // // send a request to map_loader
+  auto result{pcd_loader_client_->async_send_request(
+    request,
+    [](rclcpp::Client<autoware_map_msgs::srv::GetDifferentialPointCloudMap>::SharedFuture) {})};
+
+  std::future_status status = result.wait_for(std::chrono::seconds(0));
+  while (status != std::future_status::ready) {
+    RCLCPP_INFO(logger_, "waiting response");
+    if (!rclcpp::ok()) {
+      return;
+    }
+    status = result.wait_for(std::chrono::seconds(1));
+  }
+  update_ndt(result.get()->new_pointcloud_with_ids, result.get()->ids_to_remove);
+}
+
+void MapUpdateModule::update_ndt(
+  const std::vector<autoware_map_msgs::msg::PointCloudMapCellWithID> & maps_to_add,
+  const std::vector<std::string> & map_ids_to_remove)
+{
+  RCLCPP_INFO(
+    logger_, "Update map (Add: %d, Remove: %d)", static_cast<int>(maps_to_add.size()),
+    static_cast<int>(map_ids_to_remove.size()));
+  if (
+    (static_cast<int>(maps_to_add.size()) == 0) &
+    (static_cast<int>(map_ids_to_remove.size()) == 0)) {
```

Comment:
I think `if (maps_to_add.empty() & map_ids_to_remove.empty())` is more intuitive.

Reply:
Fixed: https://github.com/autowarefoundation/autoware.universe/pull/2339/commits/ee26d573c49dd4684e7013028ff67441ef2916fb

Reply:
Using `0UL` instead resolves the warning. Also `%ul` can be used for `RCLCPP_INFO`.

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/2339/commits/f61552f3fb079be0d44697b36820cf4c7b5c8588
Resolved